### DISCUSSION TITLE
c++ co_await

### DISCUSSION THREAD

**Discussion by unicomp21** [dotnet/runtime#70771] [Q&A] [answered]
@lewing I noticed the "tail" instruction has been implemented in wasm for blazor webassembly?  i'm wondering if the same thing could be done for c++ on webassembly to support co_await?

i'm a bit out of my depth on this one, but just know that @GorNishanov said tail calls are required for co_await.  tail calls look to be "in process" for webassembly spec, so wondering if someone has figured out a workaround in the meantime?

**Comment by GorNishanov** [comment, marked as answer]
There is a workaround for systems without tailcalls. Resume operation on the coroutine handle would become a loop.
Tail calls is used in situations when a coroutine performs symmetrical transfer, ie. coroutine 1 suspends and then resumes coroutine 2.

Here is a sketch of how such workaround might look like

```c++
void resume(coro)
{
    do {
         coro = coro.real_resume_and_return_the_coro_to_transfer_or_nullptr();
    } while(coro);
}
```

**Reply by unicomp21** [reply]
@tlively thoughts? ^ Could we do the same for emscripten?  Or is this already happening in binaryen?

**Reply by tlively** [reply]
So I just tried building the tests from https://github.com/andreasbuhr/cppcoro with Emscripten, and they did indeed fail with `error: WebAssembly 'tail-call` feature not enabled`, which means that clang is generating code with the `musttail` attribute. Any workaround for this would have to be done in the clang frontend - any later and too much information will have been lost. I filed https://discourse.llvm.org/t/supporting-coroutines-without-tail-calls/63249 to ask the clang developers about this.

**Reply by unicomp21** [reply]
@tlively i'm starting to realize maybe I have been a little blind the last couple years.  For the tests above, does co_yield pass? 
 when using webassembly/emscripten?  i'm thinking co_yield could give me 95% of what i need, if i just "shift" my brain a little.

**Reply by unicomp21** [reply]
@GorNishanov am i right in thinking co_yield should "just work" for emscripten/webassembly?

**Reply by GorNishanov** [reply]
`co_await` and `co_yield` are one and the same. They are all powered by the same underlying machinery.
However, you can write coroutine libraries that do not rely on symmetric transfer feature. If that is the case, you can simply rewrite "musttail" away.

Cppcoro has been written in the times when symmetric transfer was not available and therefore has mode where it puts workarounds at the library level.

Even if you are using most modern library, you need to use it in a certain way for the symmetric transfer to become relevant and blow up the stack.

**Reply by unicomp21** [reply]
thanks @GorNishanov !

**Reply by unicomp21** [reply]
https://github.com/dotnet/runtime/issues/70957
